File: ios/SuperheroApp/Characters/CharactersPresenter.swift
//
//  CharactersPresenter.swift
//  SuperheroApp
//
//  Created by Balazs Varga on 2018. 02. 12..
//  Copyright © 2018. W.UP. All rights reserved.
//

import Common

private let defaultLimit = Int32(100)

class CharactersPreseneter: CharactersMvpPresenter {

    private let useCaseHandler: CommonUseCaseHandler
    private var getCharactersUseCase: GetCharactersUseCase
    private weak var view: CharactersMvpView?
    private var currentPage: CommonPage = CommonPage(limit: defaultLimit, offset: 0)
    private var objects: [CommonCharacter] = [CommonCharacter]()

    init(useCaseHandler: CommonUseCaseHandler, getCharactersUseCase: GetCharactersUseCase) {
        self.useCaseHandler = useCaseHandler
        self.getCharactersUseCase = getCharactersUseCase
    }

    func takeView(view: CharactersMvpView) {
        self.view = view
        self.loadCharacters()
    }

    func characters() -> [CommonCharacter] {
        return self.objects
    }

    func charactersCount() -> Int {
        return self.objects.count
    }

    func loadCharacters() {

        self.view?.showLoadingIndicator()

        let request = GetCharactersRequest(page: self.currentPage)

        self.useCaseHandler.executeUseCase(useCase: self.getCharactersUseCase,
                                           request: request,
                                           success: { (response: CommonUseCaseResponse) in

            self.view?.hideLoadingIndicator()

            if let response = response as? GetCharactersResponse {
                if response.characters.isEmpty {
                    self.view?.showNoCharacters()
                } else {
                    self.objects.append(contentsOf: response.characters)
                    self.view?.refreshCharacters()
                }
            }
            return CommonStdlibUnit()
        }, error: {
            self.view?.hideLoadingIndicator()
            // TODO hardcoded message
            self.view?.showLoadingCharactersError(message: "Something wrong!")
            return CommonStdlibUnit()
        })
    }

    func loadMoreCharacters() {
        self.view?.showMoreLoadingIndicator()

        self.currentPage.offset += defaultLimit
        let request = GetCharactersRequest(page: self.currentPage)

        self.useCaseHandler.executeUseCase(useCase: self.getCharactersUseCase,
                                           request: request,
                                           success: { (response: CommonUseCaseResponse) in
            self.view?.hideMoreLoadingIndicator()

            if let response = response as? GetCharactersResponse {
                if response.characters.isEmpty {
                    self.view?.showNoCharacters()
                } else {
                    self.objects.append(contentsOf: response.characters)
                    self.view?.refreshCharacters()
                }
            }
            return CommonStdlibUnit()
        }, error: {
            // TODO remove hardcoded message
            self.view?.hideMoreLoadingIndicator()
            self.view?.showLoadingCharactersError(message: "Something wrong!")
            return CommonStdlibUnit()
        })
    }
}

File: common/src/main/kotlin/digital/wup/superheroapp/common/characters/CharactersPresenter.kt
package digital.wup.superheroapp.common.characters

import digital.wup.superheroapp.common.UseCaseHandler
import digital.wup.superheroapp.common.characters.domain.usecase.GetCharactersRequest
import digital.wup.superheroapp.common.characters.domain.usecase.GetCharactersResponse
import digital.wup.superheroapp.common.characters.domain.usecase.GetCharactersUseCase
import digital.wup.superheroapp.common.characters.domain.model.Page

const val DEFAULT_LIMIT = 100

class CharactersPresenter(
        private val useCaseHandler: UseCaseHandler,
        private val getCharactersUseCase: GetCharactersUseCase
) : CharactersMvpPresenter {

    private var view: CharactersMvpView? = null
    private var currentPage: Page = Page(DEFAULT_LIMIT, 0)

    override fun takeView(view: CharactersMvpView) {
        this.view = view
        this.loadCharacters()
    }

    override fun dropView() {
        view = null
    }

    override fun loadCharacters() {
        view?.showLoadingIndicator()

        val request = GetCharactersRequest(currentPage)

        useCaseHandler.executeUseCase(getCharactersUseCase, request, { response: GetCharactersResponse ->
            view?.hideLoadingIndicator()

            if (response.characters.isEmpty()) {
                view?.showNoCharacters()
            } else {
                view?.showCharacters(response.characters)
            }
        }, {
            view?.hideLoadingIndicator()
            // TODO hardcoded message
            view?.showLoadingCharactersError("Something wrong!")
        })
    }

    override fun loadMoreCharacters() {
        view?.showMoreLoadingIndicator()

        currentPage.offset += DEFAULT_LIMIT
        val request = GetCharactersRequest(currentPage)

        useCaseHandler.executeUseCase(getCharactersUseCase, request, { response: GetCharactersResponse ->
            view?.hideMoreLoadingIndicator()

                if (response.characters.isEmpty()) {
                    view?.showNoCharacters()
                } else {
                    view?.showCharacters(response.characters)
                }
        }, {
            // TODO remove hardcoded message
            view?.hideMoreLoadingIndicator()
            view?.showLoadingCharactersError("Something wrong!")
        })
    }
}